### ISSUE TITLE
For Retrieval model, can I use different set of candidates for each training sample?

### ISSUE THREAD

**Issue by jamie613** [tensorflow/recommenders#638]
I have a transaction dataset and the available items vary all the time.
Is there any way I can specify the candidates for each sample in tfrs.metrics.FactorizedTopK?

**Comment by OmarMAmin** [comment]
I believe you mean that if you'd like to change the features per candidate depending on the training sample (daily features update for candidates) right?, I was also thinking about how to do so, as for now i need to fix all the features for the candidate tower, it'd be interesting to know how this can be done.

**Comment by jamie613** [comment]
Slightly different.  My situation is that the available items at a certain time point are only a fraction of the whole item set, and so it makes more sense to select TopK from the fraction, instead of the whole item set.  I can filter the TopK at prediction stage and keep only items that are available at that time point, but it would help if I can somehow define the available items for each training sample, and have the tfrs.metrics.FactorizedTopK only evaluate on the specific available item set.

**Comment by OmarMAmin** [comment]
Aha, I had a similar situation, and actually the way we're sampling will always samples the popular items, which might not be the case you would like to optimize for, the solution for that is to add more global negatives for each batch

See this paper
https://storage.googleapis.com/pub-tools-public-publication-data/pdf/b9f4e78a8830fe5afcf2f0452862fb3c0d6584ea.pdf

It's an extension for the paper this library is built on top of, adding global uniform negatives, make your batch a bit more representative for the problem you're solving for, as when sampling uniformly, you'll have more percentage of available items, so the loss is optimizing for sth close to what will happen in production.

Sure it'd be better to have sampling per training example, but this will converge to the normal negative sampling that you can do, it'd be less efficient than the current mechanism, as for each positive pair, you'll check the query and sample from available items for that query, and you'll need to do that for the whole batch, a more efficient solution would be to just add global uniform negatives (larger than the original batch as described in the paper to make the batch as similar to what you need as possible and the training is still efficient)
